### ISSUE TITLE
Is this production-ready ?

### ISSUE THREAD

**Issue by snwfdhmp** [djrobstep/migra#168]
Really awesome tool ! Thanks for sharing !

I'm planning to use it for some weeks to automate our staging environment migrations, but then I want to apply it to production environment if no drawback is noticed.

What do you think about it ? Do you consider your tool as production-ready ?

**Comment by ChrisKitching** [comment]
As someone who does use this tool with a production database: you definitely don't want to be running the output of `migra` automatically. Migra diffs the schema and tells you how to get from A to B, but it will never be able to magically figure out what the correct data transformation to apply is.

Migra can wiggle your columns about, but it has no possible way to know how to transform the data in the database into the new format. for stuff like functions and views you can just run the `migra` script and it'll drop/recreate everything perfectly. But for data mutations, you really need to be careful.

Migra is great for partially-automating the process of doing a migration. It can also be a lifesaver for spotting mistakes as a result of having slightly messed one up. If you just run it blindly on your db as part of your CI/CD pipeline to push to production, you'll probably end up deleting half your database at some point.

**Comment by snwfdhmp** [comment]
Thank you for your feedback @ChrisKitching

**Comment by djrobstep** [comment]
This is a very good summary @ChrisKitching. I'll close this issue now, but do you mind if I add your comment as an FAQ or similar to the docs? It's solid advice.

**Comment by ChrisKitching** [comment]
> do you mind if I add your comment as an FAQ or similar to the docs? It's solid advice.

Sounds good.

One other thing to add: I found migra to be _not quite_ perfect in some advanced cases. Diffing the output of `pg_dump` can be useful for determining, after a migration, if everything has truly gone as expected. The output of such a diff is much less useful than migra for figuring out how to migrate, but it's slightly better than migra at answering the question "is everything identical?". It can be thought of as a general-purpose workaround for any current and future bugs in migra:

```bash
pg_dump -s <an empty database with the correct schema> > canon.sql
pg_dump -s  <the production database> > prod.sql

git diff --no-index --color-words prod.sql canon.sql
```

(add `-n` etc. as required).

That said, in the overwhelming majority of cases I found this produces no differences. I use approximately this in CI to determine if a migration is needed for a new version of the software.
